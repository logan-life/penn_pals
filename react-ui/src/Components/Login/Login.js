import React, { useState } from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import { useHistory } from "react-router-dom";
import ResetPassword from "./ForgotPassword";
import { BrowserRouter as Router, Link, Route, Switch } from "react-router-dom";

export default function Login() {
  const [user, setUser] = useState("");
  const [password, setPassword] = useState("");
  const history = useHistory();

  const validateForm = (e) => {
    e.preventDefault();
    const logindata = Object.freeze({
      username: user,
      password: password,
    });
    fetch("api/login", {
      method: "POST",
      body: JSON.stringify(logindata),
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((res) => {
        if (res.status === 200) {
          res.json().then((value) => {
            // The below get element Id breaks testing since the page directs on successful login
            // document.getElementById("status").innerHTML = "Login successful";
            //saving username to localStorage
            localStorage.setItem("currentUser", logindata.username);
            history.push("/profile");
          });
        } else {
          res.json().then(function (json) {
            let message = "";
            for (let index = 0; index < json.errors.length; index++) {
              message = message + "<br></br>" + json.errors[index].msg;
            }
            document.getElementById("status").innerHTML = message;
          });
          const error = new Error(res.error);
          throw error;
        }
      })
      .catch((err) => {
        console.error(err);
        document.getElementById("status").innerHTML = err;
      });
  };

  return (
    <div className="Login">
      <h1 id="status">Please login</h1>

      <div className="LoginForm">
        <Form>
          {/* ControlID sets ID on Form.Control and htmlFor on Form.Label         */}
          <Form.Group controlID="loginUser">
            <Form.Label>Username</Form.Label>
            <Form.Control
              id = "loginUser"
              type="text"
              name="user"
              placeholder="Enter username"
              value={user}
              onChange={(e) => setUser(e.target.value)}
            />
          </Form.Group>

          <Form.Group controlID="formBasicPassword">
            <Form.Label htmlFor="password">Password</Form.Label>
            <Form.Control
              id = "formBasicPassword"
              type="password"
              value={password}
              name="password"
              placeholder="Enter password"
              onChange={(e) => setPassword(e.target.value)}
            />
            <Form.Text className="text-muted">
              <Router>
                <Link id="reset" to="/ResetPassword">
                  Reset Password
                </Link>
                <Switch>
                  <Route path="/ResetPassword">
                    <ResetPassword />
                  </Route>
                </Switch>
              </Router>
            </Form.Text>
          </Form.Group>

          <br></br>
          <Button
            onClick={validateForm}
            type="submit"
            id="loginBtn"
            variant="outline-primary"
          >
            Submit
          </Button>
        </Form>
      </div>
    </div>
  );
}
