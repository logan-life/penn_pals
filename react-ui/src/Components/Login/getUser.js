export default async function getUser(passwordData) {
  try {
    const resp = await fetch("api/resetPassword", {
      method: "POST",
      body: JSON.stringify(passwordData),
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });
    return resp;
  } catch (error) {
    throw new Error(error);
  }
}
