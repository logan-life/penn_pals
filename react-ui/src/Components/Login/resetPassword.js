export default async function resetPassword(passwordData) {
  try {
    const resp = await fetch("api/resetPassword", {
      method: "PUT",
      body: JSON.stringify(passwordData),
      headers: {
        "Content-Type": "application/json",
      },
    });
    return resp;
  } catch (error) {
    throw new Error(error);
  }
}
