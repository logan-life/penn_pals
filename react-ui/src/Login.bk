.Login {
    padding: 60px 0;
    text-align: center;
  }



  form {
    max-width: 500px;
    width: 100%;
    margin: 0 auto;
  }
  
  label,
  input {
    display: block;
    width: 100%;
  }
  
  label {
    margin-bottom: 5px;
    height: 22px;
  }
  
  input {
    margin-bottom: 20px;
    padding: 10px;
    border-radius: 3px;
    border: 1px solid #777;
  }