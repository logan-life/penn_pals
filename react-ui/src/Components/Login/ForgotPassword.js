import resetPassword from "./resetPassword";
import React, { useState } from "react";
import getUser from "./getUser";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

export default function ResetPassword() {
  const [question, setQuestion] = useState("");
  const [answer, setAnswer] = useState("");
  const [password, setPassword] = useState("");
  const [user, setUser] = useState("");
  const [reset, setReset] = useState(false);
  const status = document.getElementById("status");
  async function getQuestion() {
    const userName = Object.freeze({
      user: user,
    });
    const res = await getUser(userName);
    console.log(res);
    if (res.status === 200) {
      res.json().then(function (json) {
        setQuestion(json.security_question);
      });
      setReset(true);
    } else {
      const status = document.getElementById("status");
      status.innerHTML = "This user does not exist. Please try again.";
    }
  }
  async function ResetPassword(e) {
    e.preventDefault();
    let generator = require("generate-password");
    let pwd = generator.generate({
      length: 10,
      numbers: true,
    });
    const passwordData = Object.freeze({
      password: pwd,
      answer: answer,
      user: user,
    });
    const res = await resetPassword(passwordData);
    if (res.status === 200) {
      setPassword(pwd);
      status.innerHTML =
        "Password was reset successfully.Please log in and change your password.";
    } else {
      if (res.status === 400) {
        status.innerHTML = "Answer to security question is wrong.";
        // window.alert("Answer is wrong");
      } else {
        const error = new Error(res.error);
        throw error;
      }
    }
  }

  return reset ? (
    <div id="forgotPwd">
      <div className="Profile Person">
        <h3 id="question">{question}</h3>
        <Form.Group controlID="questionForm">
          <Form.Label htmlFor="question">
            Please answer security question to reset your password:
          </Form.Label>
          <Form.Control
            type="text"
            value={answer}
            name="answer"
            id="answer"
            placeholder="Security Answer"
            onChange={(e) => setAnswer(e.target.value)}
          />
        </Form.Group>
        <label htmlFor="password">Your new password:</label>
        <br></br>
        <h4 id="question">{password}</h4>
        <Button
          onClick={ResetPassword}
          type="button"
          id="changePswd"
          variant="outline-primary"
        >
          Reset Password
        </Button>
      </div>
    </div>
  ) : (
    <div>
      <Form.Group controlID="getDataForm">
        <Form.Label htmlFor="userName">
          Please provide your user name:
        </Form.Label>
        <Form.Control
          type="text"
          value={user}
          name="name"
          id="name"
          placeholder="User Name"
          onChange={(e) => setUser(e.target.value)}
        />
      </Form.Group>
      <br></br>
      <Button
        onClick={getQuestion}
        type="button"
        id="resetPswd"
        variant="outline-primary"
      >
        Reset Password
      </Button>
    </div>
  );
}
