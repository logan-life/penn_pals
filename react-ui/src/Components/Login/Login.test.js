import React from 'react';
import Login from './Login';
import ResetPassword from './ForgotPassword';
const renderer = require('react-test-renderer');

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

// snapshot testing
test('Login view matches snapshot', () => {
const component = renderer.create(<Login />);
let tree = component.toJSON();
expect(tree).toMatchSnapshot();
});

// snapshot testing
test('ResetPassword matches snapshot', () => {
const component = renderer.create(<ResetPassword />);
let tree = component.toJSON();
expect(tree).toMatchSnapshot();
});