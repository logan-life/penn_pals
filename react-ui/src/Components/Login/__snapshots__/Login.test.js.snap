// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login view matches snapshot 1`] = `
<div
  className="Login"
>
  <h1
    id="status"
  >
    Please login
  </h1>
  <div
    className="LoginForm"
  >
    <form
      className=""
    >
      <div
        className="form-group"
        controlID="loginUser"
      >
        <label
          className="form-label"
        >
          Username
        </label>
        <input
          className="form-control"
          id="loginUser"
          name="user"
          onChange={[Function]}
          placeholder="Enter username"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        controlID="formBasicPassword"
      >
        <label
          className="form-label"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="form-control"
          id="formBasicPassword"
          name="password"
          onChange={[Function]}
          placeholder="Enter password"
          type="password"
          value=""
        />
        <small
          className="text-muted form-text"
        >
          <a
            href="/ResetPassword"
            id="reset"
            onClick={[Function]}
          >
            Reset Password
          </a>
        </small>
      </div>
      <br />
      <button
        className="btn btn-outline-primary"
        disabled={false}
        id="loginBtn"
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`ResetPassword matches snapshot 1`] = `
<div>
  <div
    className="form-group"
    controlID="getDataForm"
  >
    <label
      className="form-label"
      htmlFor="userName"
    >
      Please provide your user name:
    </label>
    <input
      className="form-control"
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="User Name"
      type="text"
      value=""
    />
  </div>
  <br />
  <button
    className="btn btn-outline-primary"
    disabled={false}
    id="resetPswd"
    onClick={[Function]}
    type="button"
  >
    Reset Password
  </button>
</div>
`;
